IdentityCore/src/util/MSIDKeyOperationUtil.m

```diff
@@ -0,0 +1,150 @@
+
+#import "MSIDKeyOperationUtil.h"
+#import "MSIDJwtAlgorithm.h"
+#import "NSData+JWT.h"
+#import "NSData+MSIDExtensions.h"
+
+@implementation MSIDKeyOperationUtil
+
++ (MSIDKeyOperationUtil *)sharedInstance
+{
+    static dispatch_once_t once;
+    static MSIDKeyOperationUtil *s_keyOperationUtil = nil;
+    dispatch_once(&once, ^{
+        s_keyOperationUtil = [[MSIDKeyOperationUtil alloc] init];
+    });
+    return s_keyOperationUtil;
+}
+
+- (BOOL)isKeyFromSecureEnclave:(SecKeyRef)key
+{
+    if (key)
+    {
+        NSDictionary *attributes = (NSDictionary *)CFBridgingRelease(SecKeyCopyAttributes(key));
+        NSString *attrTokenId = [attributes valueForKey:(__bridge NSString*)kSecAttrTokenID];
+        NSString *secureEnclaveId = (__bridge NSString*)kSecAttrTokenIDSecureEnclave;  // id = com.apple.settoken for key in secure enclave
+        return [secureEnclaveId isEqualToString:attrTokenId];
+    }
+    return NO;
+}
+
+- (BOOL)isOperationSupportedByKey:(SecKeyOperationType)operation algorithm:(SecKeyAlgorithm)algorithm key:(SecKeyRef)key context:(id<MSIDRequestContext> _Nullable)context error:(NSError * _Nullable __autoreleasing *)error
+{
+    if (key == NULL)
+    {
+        [self generateErrorWithMessage:@"Key passed in to check operation is not defined." underlyingError:nil context:context error:error];
+        return NO;
+    }
+    BOOL isSupported = SecKeyIsAlgorithmSupported(key, operation, algorithm);
+    if (!isSupported)
+    {
+        NSString *errorMessage = [NSString stringWithFormat:@"Key does not support %ld with %@", (long)operation, algorithm];
+        [self generateErrorWithMessage:errorMessage underlyingError:nil context:context error:error];
+    }
+    return isSupported;
+}
+
+- (MSIDJwtAlgorithm)getJwtAlgorithmForKey:(SecKeyRef)privateKey context:(id<MSIDRequestContext> _Nullable)context error:(NSError * _Nullable __autoreleasing * _Nullable)error
+{
+    if (privateKey == NULL)
+    {
+        [self generateErrorWithMessage:@"No key to determine signing algorithm" underlyingError:nil context:context error:error];
+        return nil;
+    }
+    
+    NSError *ecdsaAlgError;
+    if ([self isOperationSupportedByKey:kSecKeyOperationTypeSign algorithm:kSecKeyAlgorithmECDSASignatureDigestX962SHA256 key:privateKey context:context error:&ecdsaAlgError])
+    {
+        return MSID_JWT_ALG_ES256;
+    }
+
+    if ([self isOperationSupportedByKey:kSecKeyOperationTypeSign algorithm:kSecKeyAlgorithmRSASignatureDigestPKCS1v15SHA256 key:privateKey context:context error:error])
+    {
+        return MSID_JWT_ALG_RS256;
+    }
+    
+    if (error && ecdsaAlgError)
+    {
+        MSID_LOG_WITH_CTX(MSIDLogLevelError, context, @"Key does not support signing any supported algorithms or key is public key");
+    }
+    return nil;
+}
+
+- (NSData *)getSignatureForDataWithKey:(NSData *)rawData privateKey:(SecKeyRef)privateKey context:(id<MSIDRequestContext>)context error:(NSError * _Nullable __autoreleasing *)error
```

Comment:
nit: should we pass the algo as a parameter? I think in the future if you want to switch to another signing algorithm this current implementation wouldn't support

Reply:
Added algorithm as an input parameter